### PR TITLE
VIRTS-3035

### PR DESCRIPTION
**Pull request by CDJellen**
## VIRTS-3035: Pathfinder UI/UX Improvements (Part I)
Affects:  
* `app/parsers/nmap.py` 
* `app/pathfinder_gui.py`
* `app/pathfinder_svc.py`
* `scanners/nmap/scanner.py`
* `static/js/pathfinder.js`
* `templates/pathfinder.html`
Change: New functionality to

1. Set report name to file name when loading from disk,
2. Add a custom name to `nmap` scans,
3. Remove reports from memory,
4. Rename reports post-load or post-scan.

## Description

This branch adds new functionality to the current Pathfinder plugin, building off of the Pathfinder API and the new core CALDERA API v2.  Changes include adding a new view tab called "Reports" to perform a few of the missing CRUD operations, including renaming reports in memory and removing reports from memory.  The "Import Scan" functionality now captures the file name of the `.xml` or `.yml` report file, and uses this file name as the report name.  Finally, the default `nmap` scanner was expanded to include a new custom name input field, to de-clutter the report "view" screen after executing a series of reports over a live network.  

## Checklist:

- [X] My code follows the style guidelines of this project
- [X] I have performed a self-review of my own code
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works

### CODE REVIEW

**Comment by CDJellen** [comment]
Feel free to pull the branch and let me know what you think!

Areas likely pending update:
* currently the new "Reports" tab displays both the console and graph view.   Pending major changes, I am considering adding more descriptive logged status messages for the new `PATCH` and `DELETE` requests, and removing the Graph from the view screen.  I found the graph helpful in differentiating between similarly-named reports.  
* I've elected to hold off on adding dropdown help tooltips until we migrate to the new Alpine.js and Bulma framework, but can work to make this change in the current templating scheme if needed.
* There is some duplicity in being able to download reports from either the "View" or the "Reports" pane.  I've elected to keep the feature in both for now, as it makes sense to me as a user to have access to the feature either when investigating the attack graph, or in the CRUD-style menu for downloading and then removing, or renaming and then downloading.

As always, I am more than happy to make changes as needed.  Thank you!

File: static/js/pathfinder.js [hunk 1]
```diff
@@ -79,6 +85,7 @@ function restPostFile(file, callback=null, endpoint='/plugin/pathfinder/upload')
     $.ajax({
         type: 'POST',
         url: endpoint,
+
```
Comment (elegantmoose) [inline static/js/pathfinder.js:88]: errant newline?

File: static/js/pathfinder.js [hunk 2]
```diff
@@ -94,13 +101,34 @@ function restPostFile(file, callback=null, endpoint='/plugin/pathfinder/upload')
     });
 }
 
+function restDeleteFile(file, callback=null, endpoint='/plugin/pathfinder/upload'){
```
Comment (elegantmoose) [inline static/js/pathfinder.js:104]: Where is this called?

File: static/js/pathfinder.js [hunk 3]
```diff
@@ -180,8 +208,17 @@ function loadGraph(element, address){
...
+    let new_name = $('#newReportName').val();
+    console.log(current_report)
+    console.log(new_name)
+    apiV2('PATCH', '/plugin/pathfinder/api', {'index':'report','id':current_report, 'rename':new_name});
+    reloadReports();
+}
+
 function downloadVulnerabilityReport(){
-    current_report = $('#vulnerabilityReport').val();
+    current_report = $('#altVulnerabilityReport').val();
```
Comment (elegantmoose) [inline static/js/pathfinder.js:221]: What is denoted by the variable having the `alt` prefix?

**Review by elegantmoose — CHANGES_REQUESTED** [review]
First review completed.

I still need to pull down code and test but wanted to pass this along for now.

File: app/pathfinder_gui.py [hunk 4]
```diff
@@ -113,6 +119,31 @@ async def import_report(self, data):
...
+        print(f'{data}')
+        try:
+            report_id = data.get('id')
+            report = await self.data_svc.locate('vulnerabilityreports', match=dict(id=report_id))
+            report = report[0]
+            print(f'{report}')
+            report.name = data.get('rename')
+            await self.data_svc.remove('vulnerabilityreports', match=dict(id=report_id))
+            await self.data_svc.store(report)
+            return dict(status='success', output=f'report {report_id} patched')
```
[the remark this answers is not in the document]
Comment (CDJellen) [reply app/pathfinder_gui.py]: In order to use the `.display()` call, we may need to make major modifications to `pathfinder_gui` to bring it in line with the `APIv2` framework. I've removed the `output` as it was not logged in console, which brings the response in-line with the other v1 responses defied in `pathfinder_gui`.

File: static/js/pathfinder.js [hunk 5]
```diff
@@ -54,6 +59,7 @@ function importScan(){
 
 function processScan(filename){
     function processResults(data){
+        data = JSON.parse(data);
```
[the remark this answers is not in the document]
Comment (CDJellen) [reply static/js/pathfinder.js]: This call is unnecessary, and was removed per Anna's advice on the API call.  I removed this from the UI update branch, but it seems to have made its way back in.  It has been removed from the PR branch as outdated.  Thanks!

File: static/js/pathfinder.js (hunk 3 above)
Comment (CDJellen) [reply static/js/pathfinder.js]: The `reports` tab was removed, as is caused feature duplication and created an issue with the jQuery dropdown of vulnerability reports.  This id was removed.  Thanks for the input, I'm much happier with keeping all the CRUD operations in one place.

File: app/pathfinder_gui.py [hunk 6]
```diff
@@ -93,9 +98,10 @@ async def pathfinder_core(self, request):
 
     async def scan(self, data):
         scanner = data.pop('scanner', None)
-        filename = sanitize_filename('pathfinder_%s' % date.today().strftime("%b-%d-%Y"))
-        report_file = '%s/reports/%s.xml' % (settings.data_dir, filename)
         fields = data.pop('fields', None)
+        filename = fields.pop('filename') or sanitize_filename('pathfinder_%s' % date.today().strftime("%b-%d-%Y"))
```
Comment (elegantmoose) [inline app/pathfinder_gui.py:102]: change to f-strings

File: app/pathfinder_svc.py [hunk 7]
```diff
@@ -29,8 +31,8 @@ async def import_scan(self, scan_format, report):
             temp_file = '%s/_temp_report_file.tmp' % settings.data_dir
             with open(temp_file, 'wb') as f:
                 f.write(contents)
-            parsed_report = self.parsers[scan_format].parse(temp_file)
-            parsed_report = self.enrich_report(parsed_report)
+            parsed_report = self.parsers[scan_format].parse(temp_file, report)
+            #parsed_report = self.enrich_report(parsed_report)
```
Comment (elegantmoose) [inline app/pathfinder_svc.py:35]: stale? or do we actually want to keep this?

**Review by elegantmoose — CHANGES_REQUESTED** [review]

**Comment by CDJellen** [comment]
I really like that change, I'll make the edit shortly.

File: app/pathfinder_svc.py (hunk 7 above)
Comment (CDJellen) [reply app/pathfinder_svc.py:35]: The report enrichment is feature is still in development, I've made these changes with the NMAP scan and import features in mind, which do not leverage enrichment.  Pathfinder uses a single method for report import/parsing, but we could differentiate this in the future.

**Comment by elegantmoose** [comment]
This will close #40

**Review by elegantmoose — APPROVED** [review]

**Pull request merged**